### PR TITLE
Additional class-based factories for discussion-related tests

### PR DESCRIPTION
**Pull request by notbakaneko**
Convert `BeatmapDiscussionPost` to use class-based factory and add some extra conveniences when factorying `User`s

### CODE REVIEW

File: database/factories/UserFactory.php [hunk 1]
```diff
@@ -115,4 +115,13 @@ public function withNote()
     {
         return $this->has(UserAccountHistory::factory(), 'accountHistories');
     }
+
+    public function withPlays(?int $count = null, ?string $mode = 'osu')
+    {
+        $sMode = studly_case($mode);
+
+        return $this->has("App\\Models\\UserStatistics\\{$sMode}"::factory()->state([
```
Comment (nanaya) [inline database/factories/UserFactory.php:123]: `App\Models\UserStatistics\Model::getClass($mode)` is a thing

File: tests/BeatmapsetDisqualifyNotificationsTest.php [hunk 2]
```diff
@@ -144,10 +144,7 @@ protected function setUp(): void
         Event::fake();
 
         $owner = User::factory()->create();
```
Comment (nanaya) [inline tests/BeatmapsetDisqualifyNotificationsTest.php:146]: not used

File: database/factories/BeatmapsetFactory.php [hunk 3]
```diff
@@ -57,6 +59,12 @@ public function noDiscussion()
         return $this->state(['discussion_enabled' => false]);
     }
 
+    public function owner(?User $user = null)
+    {
+        return $this->state(['user_id' => $user ?? User::factory()])
+            ->afterMaking(fn (Beatmapset $beatmapset) => $beatmapset->creator = $beatmapset->user->username);
```
Comment (nanaya) [inline database/factories/BeatmapsetFactory.php:65]: update `creator` attribute function in definition instead (put at the bottom so `user_id` is set first).

**Review by nanaya — CHANGES_REQUESTED** [review]

File: database/factories/BeatmapsetFactory.php (hunk 3 above)
Comment (notbakaneko) [reply database/factories/BeatmapsetFactory.php:65]: `user` isn't part of attributes available in `definition`?

Comment (nanaya) [reply database/factories/BeatmapsetFactory.php:65]: it's not but can be looked up `User::find($attr['user_id'])->.. ?? ...`

Comment (notbakaneko) [reply database/factories/BeatmapsetFactory.php:65]: or it can be set without another lookup....

Comment (nanaya) [reply database/factories/BeatmapsetFactory.php:65]: Even then this should be part of configuration and not just one specific state (and remove default `creator` in that case).

File: database/factories/BeatmapsetFactory.php [hunk 4]
```diff
@@ -10,16 +10,21 @@
...
+use App\Models\User;
 use Illuminate\Database\Eloquent\Factories\Factory;
 
 class BeatmapsetFactory extends Factory
 {
     protected $model = Beatmapset::class;
 
+    public function configure()
+    {
+        return $this->afterMaking(fn (Beatmapset $beatmapset) => $beatmapset->creator = $beatmapset->user?->username ?? $this->faker->userName());
```
Comment (nanaya) [inline database/factories/BeatmapsetFactory.php:22]: I think setting with `??=` is more appropriate here

(also the line is getting kinda long)

**Review by nanaya — CHANGES_REQUESTED** [review]

**Review by nanaya — APPROVED** [review]

**Pull request merged**